Sources/gattserver/DarwinLocation.swift

```diff
@@ -32,48 +48,171 @@ public class DarwinLocation: NSObject, LocationManager {
     
     // MARK: - Initialization
     
-    public override init() {
+    public init(didUpdate: ((Location) -> ())? = nil) throws {
         
+        // initialize properties
         self.internalManager = CLLocationManager()
+        self.delegate = InternalDelegate()
+        self.didUpdate = didUpdate
+        
+        // set delegate
+        self.delegate.locationManager = self
+        internalManager.delegate = self.delegate
         
-        super.init()
+        // start updating location
+        try start()
+    }
+    
+    deinit {
         
-        internalManager.delegate = self
+        stop()
     }
     
     // MARK: - Methods
     
-    public func start() throws {
+    // only call once
+    private func start() throws {
+        
+        let semaphore = Semaphore(timeout: 30, operation: .startUpdatingLocation)
+        accessQueue.sync { [unowned self] in self.internalState.start.semaphore = semaphore }
+        defer { accessQueue.sync { [unowned self] in self.internalState.start.semaphore = nil } }
         
         internalManager.startUpdatingLocation()
+        
+        try semaphore.wait()
     }
     
-    public func stop() {
+    private func stop() {
         
         internalManager.stopUpdatingLocation()
     }
 }
 
 // MARK: - CLLocationManagerDelegate
 
-extension DarwinLocation: CLLocationManagerDelegate {
+extension DarwinLocationManager {
+    
+    @objc(DarwinLocationManagerInternalDelegate)
+    final class InternalDelegate: NSObject {
+        
+        weak var locationManager: DarwinLocationManager?
+    }
+}
+
+extension DarwinLocationManager.InternalDelegate: CLLocationManagerDelegate {
     
+    @objc
     public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
         
         /**
          An array of CLLocation objects containing the location data. This array always contains at least one object representing the current location. If updates were deferred or if multiple locations arrived before they could be delivered, the array may contain additional entries. The objects in the array are organized in the order in which they occurred. Therefore, the most recent location update is at the end of the array.
          */
         
         guard let location = locations.last
-            else { return }
+            else { assertionFailure("Array always contains at least one object representing the current location."); return }
         
-        didUpdateLocation?(location.coordinate.latitude, location.coordinate.longitude)
+        self.locationManager?.accessQueue.async { [weak self] in
```

Comment:
Don't you need to use here `self.locationManager?.accessQueue.async(flags: .barrier)`?

Reply:
sure, improve as you see fit.